Reduce the size of bmalloc_heap_inlines.h

smfr opened the pull request:
#### da8edba830368030b9b07745629991509299ee4c
<pre>
Reduce the size of bmalloc_heap_inlines.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=283926">https://bugs.webkit.org/show_bug.cgi?id=283926</a>
<a href="https://rdar.apple.com/140803289">rdar://140803289</a>

Reviewed by Mark Lam.

bmalloc_heap_inlines.h is the single most expensive header in WebKit, taking about 2.3s to parse in every compilation
unit which includes it (which is most of them).

bmalloc_heap_inlines.h also has some functions declared in it which are internal to bmalloc, so move those to
a new bmalloc_heap_internal.h header, which is included from the few bmalloc cpp files that need it.

Also remove some includes from IsoHeapInlines.h which don&apos;t seem to be required there.

Parsing bmalloc_heap_inlines.h goes from 2.3s to 1.5s with this change.

* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap.c:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h:
(bmalloc_try_iso_allocate_inline): Deleted.
(bmalloc_iso_allocate_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_size_inline): Deleted.
(bmalloc_try_iso_allocate_zeroed_array_by_size_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_size_with_alignment_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_count_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_count_with_alignment_inline): Deleted.
(bmalloc_iso_allocate_array_by_size_inline): Deleted.
(bmalloc_iso_allocate_zeroed_array_by_size_inline): Deleted.
(bmalloc_iso_allocate_array_by_size_with_alignment_inline): Deleted.
(bmalloc_iso_allocate_array_by_count_inline): Deleted.
(bmalloc_iso_allocate_array_by_count_with_alignment_inline): Deleted.
(bmalloc_try_iso_reallocate_array_by_size_inline): Deleted.
(bmalloc_iso_reallocate_array_by_size_inline): Deleted.
(bmalloc_try_iso_reallocate_array_by_count_inline): Deleted.
(bmalloc_iso_reallocate_array_by_count_inline): Deleted.
(bmalloc_try_allocate_flex_inline): Deleted.
(bmalloc_allocate_flex_inline): Deleted.
(bmalloc_try_allocate_zeroed_flex_inline): Deleted.
(bmalloc_allocate_zeroed_flex_inline): Deleted.
(bmalloc_try_allocate_flex_with_alignment_inline): Deleted.
(bmalloc_allocate_flex_with_alignment_inline): Deleted.
(bmalloc_try_reallocate_flex_inline): Deleted.
(bmalloc_reallocate_flex_inline): Deleted.
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_innards.h:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_internal.h: Copied from Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h.
(bmalloc_try_iso_allocate_inline):
(bmalloc_iso_allocate_inline):
(bmalloc_try_iso_allocate_array_by_size_inline):
(bmalloc_try_iso_allocate_zeroed_array_by_size_inline):
(bmalloc_try_iso_allocate_array_by_size_with_alignment_inline):
(bmalloc_try_iso_allocate_array_by_count_inline):
(bmalloc_try_iso_allocate_array_by_count_with_alignment_inline):
(bmalloc_iso_allocate_array_by_size_inline):
(bmalloc_iso_allocate_zeroed_array_by_size_inline):
(bmalloc_iso_allocate_array_by_size_with_alignment_inline):
(bmalloc_iso_allocate_array_by_count_inline):
(bmalloc_iso_allocate_array_by_count_with_alignment_inline):
(bmalloc_try_iso_reallocate_array_by_size_inline):
(bmalloc_iso_reallocate_array_by_size_inline):
(bmalloc_try_iso_reallocate_array_by_count_inline):
(bmalloc_iso_reallocate_array_by_count_inline):
(bmalloc_try_allocate_flex_inline):
(bmalloc_allocate_flex_inline):
(bmalloc_try_allocate_zeroed_flex_inline):
(bmalloc_allocate_zeroed_flex_inline):
(bmalloc_try_allocate_flex_with_alignment_inline):
(bmalloc_allocate_flex_with_alignment_inline):
(bmalloc_try_reallocate_flex_inline):
(bmalloc_reallocate_flex_inline):

Canonical link: <a href="https://commits.webkit.org/287235@main">https://commits.webkit.org/287235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/824b1e619ef1674d59fa5ba5a70a96f36d323d18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78877 "39 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57920 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83535 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30136 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67069 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6201 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61775 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19697 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81944 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51810 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71766 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42078 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25969 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28479 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72017 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70274 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84907 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78106 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6241 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69997 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67803 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69248 "Found 3 new API test failures: /TestWebKit:WebKit.EvaluateJavaScriptThatThrowsAnException, /TestWebKit:WebKit2UserMessageRoundTripTest.WKURLRequestRef, /TestWebKit:WebKit.GeolocationBasic (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17245 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13297 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12035 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100419 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6186 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21928 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6167 "Built successfully") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9606 "Hash 824b1e61 for PR 37328 does not build (failure)") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/bmalloc/IsoHeapInlines.h:
```diff
@@ -26,10 +26,6 @@
 #pragma once
 
 #include "BPlatform.h"
-#include "DeferredDecommitInlines.h"
-#include "DeferredTriggerInlines.h"
-#include "EligibilityResultInlines.h"
-#include "FreeListInlines.h"
```
Are you sure that these are not needed, and not just due to these having being included by other files already?  I mean, is this going to break non-unified builds?

smfr replied:
I undid this change to be conservative (but this is indicative of how bmalloc internals are leaking out).

MenloDorian commented on Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h:
```diff
@@ -23,8 +23,7 @@
 
-#ifndef BMALLOC_HEAP_INLINES_H
-#define BMALLOC_HEAP_INLINES_H
+#pragma once
```
Please revert this.  bmalloc is meant to be a C project that can be #included in C files, and we would like to retain that until we have a strong reason not to.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h:
```diff
@@ -597,6 +315,3 @@ PAS_END_EXTERN_C;
 #endif /* PAS_ENABLE_BMALLOC */
 
 PAS_IGNORE_WARNINGS_END
-
-#endif /* BMALLOC_HEAP_INLINES_H */
-
```
Ditto: please revert.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/bmalloc_heap_innards.h:
```diff
@@ -23,8 +23,7 @@
 
-#ifndef BMALLOC_HEAP_INNARDS_H
-#define BMALLOC_HEAP_INNARDS_H
+#pragma once
```
Ditto: please revert.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/bmalloc_heap_innards.h:
```diff
@@ -45,6 +44,3 @@ PAS_API extern pas_allocator_counts bmalloc_allocator_counts;
 PAS_END_EXTERN_C;
 
 #endif /* PAS_ENABLE_BMALLOC */
-
-#endif /* BMALLOC_HEAP_INNARDS_h */
-
```
Ditto: please revert.

MenloDorian reviewed (approved):
r=me with fixes.

smfr committed:
Reduce the size of bmalloc_heap_inlines.h
https://bugs.webkit.org/show_bug.cgi?id=283926
rdar://140803289

Reviewed by Mark Lam.

bmalloc_heap_inlines.h is the single most expensive header in WebKit, taking about 2.3s to parse in every compilation
unit which includes it (which is most of them).

bmalloc_heap_inlines.h also has some functions declared in it which are internal to bmalloc, so move those to
a new bmalloc_heap_internal.h header, which is included from the few bmalloc cpp files that need it.

Also remove some includes from IsoHeapInlines.h which don't seem to be required there.

Parsing bmalloc_heap_inlines.h goes from 2.3s to 1.5s with this change.

* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/IsoHeap.cpp:
* Source/bmalloc/bmalloc/IsoHeapInlines.h:
* Source/bmalloc/bmalloc/TZoneHeap.cpp:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap.c:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h:
(bmalloc_try_iso_allocate_inline): Deleted.
(bmalloc_iso_allocate_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_size_inline): Deleted.
(bmalloc_try_iso_allocate_zeroed_array_by_size_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_size_with_alignment_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_count_inline): Deleted.
(bmalloc_try_iso_allocate_array_by_count_with_alignment_inline): Deleted.
(bmalloc_iso_allocate_array_by_size_inline): Deleted.
(bmalloc_iso_allocate_zeroed_array_by_size_inline): Deleted.
(bmalloc_iso_allocate_array_by_size_with_alignment_inline): Deleted.
(bmalloc_iso_allocate_array_by_count_inline): Deleted.
(bmalloc_iso_allocate_array_by_count_with_alignment_inline): Deleted.
(bmalloc_try_iso_reallocate_array_by_size_inline): Deleted.
(bmalloc_iso_reallocate_array_by_size_inline): Deleted.
(bmalloc_try_iso_reallocate_array_by_count_inline): Deleted.
(bmalloc_iso_reallocate_array_by_count_inline): Deleted.
(bmalloc_try_allocate_flex_inline): Deleted.
(bmalloc_allocate_flex_inline): Deleted.
(bmalloc_try_allocate_zeroed_flex_inline): Deleted.
(bmalloc_allocate_zeroed_flex_inline): Deleted.
(bmalloc_try_allocate_flex_with_alignment_inline): Deleted.
(bmalloc_allocate_flex_with_alignment_inline): Deleted.
(bmalloc_try_reallocate_flex_inline): Deleted.
(bmalloc_reallocate_flex_inline): Deleted.
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_innards.h:
* Source/bmalloc/libpas/src/libpas/bmalloc_heap_internal.h: Copied from Source/bmalloc/libpas/src/libpas/bmalloc_heap_inlines.h.
(bmalloc_try_iso_allocate_inline):
(bmalloc_iso_allocate_inline):
(bmalloc_try_iso_allocate_array_by_size_inline):
(bmalloc_try_iso_allocate_zeroed_array_by_size_inline):
(bmalloc_try_iso_allocate_array_by_size_with_alignment_inline):
(bmalloc_try_iso_allocate_array_by_count_inline):
(bmalloc_try_iso_allocate_array_by_count_with_alignment_inline):
(bmalloc_iso_allocate_array_by_size_inline):
(bmalloc_iso_allocate_zeroed_array_by_size_inline):
(bmalloc_iso_allocate_array_by_size_with_alignment_inline):
(bmalloc_iso_allocate_array_by_count_inline):
(bmalloc_iso_allocate_array_by_count_with_alignment_inline):
(bmalloc_try_iso_reallocate_array_by_size_inline):
(bmalloc_iso_reallocate_array_by_size_inline):
(bmalloc_try_iso_reallocate_array_by_count_inline):
(bmalloc_iso_reallocate_array_by_count_inline):
(bmalloc_try_allocate_flex_inline):
(bmalloc_allocate_flex_inline):
(bmalloc_try_allocate_zeroed_flex_inline):
(bmalloc_allocate_zeroed_flex_inline):
(bmalloc_try_allocate_flex_with_alignment_inline):
(bmalloc_allocate_flex_with_alignment_inline):
(bmalloc_try_reallocate_flex_inline):
(bmalloc_reallocate_flex_inline):

Canonical link: https://commits.webkit.org/287235@main
